### PR TITLE
Really clean-up containers and images a-la-Docker

### PR DESCRIPTION
**Pull request by alessio**

### CODE REVIEW

File: dokku [hunk 1]
```diff
@@ -122,9 +122,9 @@ case "$1" in
 
   cleanup)
     # delete all non-running container
-    docker ps -a | grep 'Exit' | awk '{print $1}' | xargs docker rm &> /dev/null &
+    docker rm $(docker ps -a -f "status=exited" -q) &> /dev/null &
     # delete unused images
-    docker images | grep '<none>' | awk '{print $3}' | xargs docker rmi &> /dev/null &
+    docker rmi $(sudo docker images -f "dangling=true" -q) &> /dev/null &
```
Comment (vincentfretin) [inline dokku]: why the sudo here? I think you can remove it. 
Be careful, when there is no image to delete, you will have an error (not shown here, be the exit status will not be 0)
I use this in one of my script:

# remove exited containers
exited_containers=$(docker ps -a -q -f status=exited)
[[ -z "$exited_containers" ]] || docker rm $exited_containers &> /dev/null &

# remove dangling images
dangling_images=$(docker images -q -f dangling=true)
[[ -z "$dangling_images" ]] || docker rmi $dangling_images &> /dev/null &

Comment (alessio) [inline dokku]: Fixed the sudo thing, was just cruft leftover.

> Be careful, when there is no image to delete, you will have an error

The same would happen with the old version too. BTW, your 2-liners look fine to me.

File: dokku [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+7#!/usr/bin/env bash
```
Comment (michaelshobbs) [inline dokku]: Typo?

**Comment by michaelshobbs** [comment]
Would be great to have a unit test for cleanup.

File: dokku (hunk 2 above)
Comment (alessio) [inline dokku]: On Tue, Feb 10, 2015 at 2:14 PM, Michael Hobbs <notifications@github.com> wrote:
>
> In dokku:
>
> > @@ -1,4 +1,4 @@
> > -#!/usr/bin/env bash
> > +7#!/usr/bin/env bash
>
> Typo?
>

Yep



-- 
Alessio Treglia          | www.alessiotreglia.com
Debian Developer         |     alessio@debian.org
Ubuntu Core Developer    |  quadrispro@ubuntu.com
0416 0004 A827 6E40 BB98 90FB E8A4 8AE5 311D 765A

**Comment by josegonzalez** [comment]
@alessio bump for unit test :)

**Comment by josegonzalez** [comment]
@michaelshobbs Do these tests look good to you?

**Comment by josegonzalez** [comment]
Hmm I wonder why tests didn't trigger here...

**Comment by michaelshobbs** [comment]
:+1: on the tests. I think @alessio will need to follow his fork on CircleCI to make the tests run.

**Comment by josegonzalez** [comment]
@progrium Can you turn on pull request support for dokku? This is the message I got from CircleCI support:

> We don't enable pull requests from forks by default because that can sometimes introduce the risk of exposing private credentials depending on how people use Circle. You can allow PRs from forks to be built by turning on "Project fork pull requests" here: https://circleci.com/gh/progrium/dokku/edit#experimental

**Comment by progrium** [comment]
Mmmmm... I'd really like to avoid it. Might want to put secrets on CircleCI
and this would let anybody get them.

On Thu, Feb 19, 2015 at 12:52 PM, Jose Diaz-Gonzalez <
notifications@github.com> wrote:

> @progrium <https://github.com/progrium> Can you turn on pull request
> support for dokku? This is the message I got from CircleCI support:
>
> We don't enable pull requests from forks by default because that can
> sometimes introduce the risk of exposing private credentials depending on
> how people use Circle. You can allow PRs from forks to be built by turning
> on "Project fork pull requests" here:
> https://circleci.com/gh/progrium/dokku/edit#experimental
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/progrium/dokku/pull/967#issuecomment-75112159>.
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by josegonzalez** [comment]
We don't currently have secrets - feel free to correct me @michaelshobbs - and I don't think our testing infra needs them anymore (the old travis stuff did).

We can always disable if we ever do.

**Comment by progrium** [comment]
I can think of a time in the not so distant future when we might. And I
don't want to accidentally forget to disable. But I guess that's a hard
case to make you care about since it would be on me anyway. Did we have PR
testing before?

On Thu, Feb 19, 2015 at 1:23 PM, Jose Diaz-Gonzalez <
notifications@github.com> wrote:

> We don't currently have secrets - feel free to correct me @michaelshobbs
> <https://github.com/michaelshobbs> - and I don't think our testing infra
> needs them anymore (the old travis stuff did).
>
> We can always disable if we ever do.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/progrium/dokku/pull/967#issuecomment-75117904>.
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by josegonzalez** [comment]
I guess it was just PRs from @michaelshobbs and myself. 

Maybe its just me, but the value of having contributors write tests decrease if they don't get those tests run. And having them turn on an extra knob increases friction to contributing, so I'm torn.

Is there an example where we'd want to have secrets in our circleci setup?

**Pull request merged**